go/libraries/doltcore/sqle/statspro/stats_provider.go

```diff
@@ -336,3 +352,108 @@ func (p *Provider) DataLength(ctx *sql.Context, db string, table sql.Table) (uin
 
 	return priStats.AvgSize(), nil
 }
+
+func (p *Provider) Prune(ctx *sql.Context) error {
+	dSess := dsess.DSessFromSess(ctx.Session)
+
+	for _, sqlDb := range p.pro.DoltDatabases() {
+		dbName := strings.ToLower(sqlDb.Name())
+		sqlDb, ok, err := dSess.Provider().SessionDatabase(ctx, dbName)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			continue
+		}
+		statDb, ok := p.getStatDb(dbName)
+		if !ok {
+			continue
+		}
+
+		p.CancelRefreshThread(dbName)
+
+		tables, err := sqlDb.GetTableNames(ctx)
+		if err != nil {
+			return err
+		}
+
+		for _, branch := range statDb.Branches() {
+			var stats []sql.Statistic
+			for _, t := range tables {
+				tableStats, err := p.GetTableDoltStats(ctx, branch, dbName, t)
+				if err != nil {
+					return err
+				}
+				stats = append(stats, tableStats...)
+			}
+
+			if err := p.DropDbBranchStats(ctx, branch, dbName, true); err != nil {
+				return err
+			}
+
+			for _, s := range stats {
+				ds, ok := s.(*DoltStats)
+				if !ok {
+					return fmt.Errorf("unexpected statistics type found: %T", s)
+				}
+				if err := statDb.SetStat(ctx, branch, ds.Qualifier(), ds); err != nil {
+					return err
+				}
+			}
+			if err := statDb.Flush(ctx, branch); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (p *Provider) Purge(ctx *sql.Context) error {
```

Comment:
Seems like both oft these need some kind of locking to prevent another session from doing an update at the same time

Reply:
There is probably a race where someone calls analyze, purge, and the analyze lands after the purge. this is the same race that exists elsewhere for other stats put/delete. I'll take a pass at having analyze and auto update check the context before doing any puts -- delete operations cancel the context and i think that should narrow the window for races.